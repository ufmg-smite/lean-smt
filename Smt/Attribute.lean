import Lean
import Std

namespace Smt.Attribute

open Lean
open Std

/-- An extension to Lean's runtime environment to support SMT attributes.
    Maintains a set of function declarations for the `smt` tactic to utilize
    while generating the SMT query. -/
abbrev SmtExtension := SimpleScopedEnvExtension Name (HashSet Name)

/-- Adds a declaration to the set of function declarations maintained by the SMT
    environment extension. -/
def addSmtEntry (d : HashSet Name) (e : Name) : HashSet Name :=
  d.insert e

initialize smtExt : SmtExtension ← registerSimpleScopedEnvExtension {
  name     := `SmtExt
  initial  := {}
  addEntry := addSmtEntry
}

/-- Throws unexpected type error. -/
def throwUnexpectedType (t : Name) (n : Name) : AttrM Unit :=
  throwError s!"unexpected type at '{n}', `{t}` expected"

/-- Validates the tagged declaration. -/
def validate (n : Name) : AttrM Unit := do
  match (← getEnv).find? n with
  | none      => throwError s!"unknown constant '{n}'"
  | some info =>
    match info.type with
    | Expr.const c .. =>
      if c != (`Smt.Transformer) then throwUnexpectedType `Smt.Transformer n
    | _               => throwUnexpectedType `Smt.Transformer n

/-- Registers an SMT attribute with the provided name and description and links
    it against `ext`. -/
def registerSmtAttr (attrName : Name) (attrDescr : String)
  : IO Unit :=
  registerBuiltinAttribute {
    name  := attrName
    descr := attrDescr
    applicationTime := AttributeApplicationTime.afterTypeChecking
    add   := fun decl stx attrKind => do
      trace[Smt.debug.attr] s!"attrName: {attrName}, attrDescr: {attrDescr}"
      trace[Smt.debug.attr] s!"decl: {decl}, stx: {stx}, attrKind: {attrKind}"
      Attribute.Builtin.ensureNoArgs stx
      validate decl
      setEnv (smtExt.addEntry (← getEnv) decl)
      trace[Smt.debug.attr]
        s!"transformers: {(smtExt.getState (← getEnv)).toList}"
    erase := fun declName => do
      let s := smtExt.getState (← getEnv)
      let s := s.erase declName
      modifyEnv fun env => smtExt.modifyState env fun _ => s
  }

initialize registerSmtAttr `Smt
             "Utilize this function to transform Lean expressions to SMT terms."

end Smt.Attribute
