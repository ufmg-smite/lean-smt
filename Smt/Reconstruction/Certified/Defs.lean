import Smt.Reconstruction.Env
import Smt.Reconstruction.Utils
import Smt.Reconstruction.Types
import Smt.Reconstruction.Coe
import Smt.Reconstruction.Term

open Types
open proof
open term
open sort

set_option maxHeartbeats 500000

@[simp] def combineProps : Interpretation → Interpretation → (Prop → Prop → Prop) → Interpretation
| some ⟨ boolSort, k₁ ⟩, some ⟨ boolSort, k₂ ⟩, f =>
    some ⟨ boolSort, λ Γ Δ => f (k₁ Γ Δ) (k₂ Γ Δ) ⟩ -- maybe we can use state monad to avoid propagating the environments manually?
| _, _, _ => none

@[simp] def interpTerm : term → Interpretation
| term.const   n  s  => some ⟨ s, λ Γ Δ => Γ n Δ s  ⟩
| term.not     t₁    => match interpTerm t₁ with
                        | some ⟨ boolSort, k ⟩ => some ⟨ boolSort, λ Γ Δ => ¬ (k Γ Δ) ⟩
                        | _ => none
| term.and     t₁ t₂ => combineProps (interpTerm t₁) (interpTerm t₂) (λ p₁ p₂ => p₁ ∧ p₂)
| term.or      t₁ t₂ => combineProps (interpTerm t₁) (interpTerm t₂) (λ p₁ p₂ => p₁ ∨ p₂)
| term.implies t₁ t₂ => combineProps (interpTerm t₁) (interpTerm t₂) (λ p₁ p₂ => p₁ → p₂)
| term.xor     t₁ t₂ => combineProps (interpTerm t₁) (interpTerm t₂) (λ p₁ p₂ => p₁ ≠ p₂)
| term.eq      t₁ t₂ => match interpTerm t₁, interpTerm t₂ with
                        | some ⟨ s₁, k₁ ⟩, some ⟨ s₂, k₂ ⟩ =>
                          if r: s₂ = s₁
                          then some ⟨ boolSort, λ Γ Δ =>
                                            let coercion := coe' (congrArg (interpSort Δ) r) 
                                            k₁ Γ Δ = coercion (k₂ Γ Δ) ⟩
                          else none
                        | _, _ => none
| term.bot           => some ⟨ boolSort, λ _ _ => False ⟩
| term.top           => some ⟨ boolSort, λ _ _ => True  ⟩
| term.app     t₁ t₂ => match interpTerm t₁, interpTerm t₂ with
                        | some ⟨ arrow s₁₁ s₁₂, k₁ ⟩, some ⟨ s₂, k₂ ⟩ =>
                          if r: s₂ = s₁₁
                          then some ⟨ s₁₂ , λ Γ Δ =>
                                              let coercion := coe' (congrArg (interpSort Δ) r)
                                              k₁ Γ Δ (coercion (k₂ Γ Δ)) ⟩
                          else none
                        | _, _ => none
| _ => none

@[simp] def validWith (Γ : Environment) (Δ : SEnvironment) (t : term) : Prop :=
  match interpTerm t with
  | some ⟨ boolSort, k ⟩ => Coe.coe (k Γ Δ)
  | _ => False

def followsFrom (t₁ t₂ : term) : Prop :=
  ∀ {Γ : Environment} {Δ : SEnvironment}, validWith Γ Δ t₁ → validWith Γ Δ t₂

@[simp] def wellTyped (t : term) : Prop := isSome (interpTerm t)

@[simp] def getInterp (t : term) (h : wellTyped t) : Σ (s : sort), Environment → (Δ : SEnvironment) → interpSort Δ s :=
  match r: interpTerm t with
  | some ⟨ s, k ⟩ => ⟨ s, k ⟩
  | none => by simp at h; rewrite [r] at h; exact (False.elim h)

