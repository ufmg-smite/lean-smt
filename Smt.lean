import Smt.General
import Smt.Nat
import Smt.Prop
import Smt.Solver
import Smt.String
import Smt.Tactic
import Smt.Term
