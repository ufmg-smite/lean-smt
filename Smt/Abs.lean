/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Translator
import Mathlib.Algebra.Order.Group.Abs

namespace Smt.Abs

open Lean Expr
open Translator Term


@[smtTranslator] def replaceAbs : Translator
  | app (app (app (app (const ``abs _) _) _) _) e => do
      let tmE ← applyTranslators! e
      return appT (symbolT "abs") tmE
  | _ => return none

end Smt.Abs
