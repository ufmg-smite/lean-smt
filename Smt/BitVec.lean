/-
Copyright (c) 2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Wojciech Nawrocki
-/

import Lean
import Smt.Translator
import Smt.Data.BitVec

namespace Smt.BitVec

open Lean Expr
open Translator Term

@[smtTranslator] def replaceType : Translator
  | e@(app (const ``BitVec _) n) => do
    let n ← Meta.whnf n
    let some n ← Meta.evalNat n |>.run
      | throwError "bitvector type{indentD e}\nhas variable width"
    return mkApp2 (symbolT "_") (symbolT "BitVec") (literalT (toString n))
  | _ => return none

def mkLit (w : Nat) (n : Nat) : Term :=
  let bits := Nat.toDigits 2 n |>.take w
  literalT <| bits.foldl (init := "#b".pushn '0' (w - bits.length)) (·.push ·)

@[smtTranslator] def replaceEq : Translator
  -- TODO: we should really just support beq/bne across all types uniformly
  | app (app (const ``BEq.beq _) (app (const ``BitVec _) _)) _ => return symbolT "="
  | app (app (app (app (const ``bne _) (app (const ``BitVec _) _)) _) e) e' =>
    return appT (symbolT "not") (mkApp2 (symbolT "=") (← applyTranslators! e) (← applyTranslators! e'))
  | _ => return none

@[smtTranslator] def replaceFun : Translator
  | app (app (const ``BitVec.append _) _) _ => return symbolT "concat"
  | app (const ``BitVec.and _) _            => return symbolT "bvand"
  | app (const ``BitVec.or _) _             => return symbolT "bvor"
  | app (const ``BitVec.xor _) _            => return symbolT "bvxor"
  | e@(app (app (app (const ``BitVec.shiftLeft _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvshl") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.shiftRight _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvlshr") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.extract _) _) i) j) => do
    let i ← reduceLit i e
    let j ← reduceLit j e
    return mkApp3 (symbolT "_") (symbolT "extract") (literalT (toString i)) (literalT (toString j))
  | e@(app (const ``BitVec.zero _) w) => do
    let w ← reduceWidth w e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w 0
  | e@(app (app (const ``BitVec.ofNat _) w) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w n
  | _ => return none
where
  reduceWidth (w : Expr) (e : Expr) : TranslationM Nat := do
    let some w ← Meta.evalNat (← Meta.whnf w) |>.run
      | throwError "bitvector width{indentD w}\nis not constant in{indentD e}"
    return w
  reduceLit (n : Expr) (e : Expr) : TranslationM Nat := do
    let some n ← Meta.evalNat (← Meta.whnf n) |>.run
      | throwError "literal{indentD n}\nis not constant in{indentD e}"
    return n

end Smt.BitVec
