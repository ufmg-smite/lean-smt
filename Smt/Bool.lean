/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Transformer

namespace Smt.Bool

open Lean Expr
open Smt.Transformer

@[Smt] def replaceConst : Transformer
  | const `Bool.true ..                                 => pure (mkConst `true)
  | const `Bool.false ..                                => pure (mkConst `false)
  | app (app (const `BEq.beq ..) (const `Bool ..) _) .. =>
    pure (mkConst (Name.mkSimple "="))
  | e                                                   => pure e

@[Smt] def replaceEq : Transformer
  | app (app (const `Decidable.decide ..)
             (app (app (app (const `Eq ..) (const `Bool ..) _)
                       a _) b _) ..) .. =>
    return match ← applyTransformations a, ← applyTransformations b with
    | some a', some b' =>
      mkApp2 (mkConst (Name.mkSimple "=")) a' b'
    | _      , _       => none
  | e                                   => pure e

end Smt.Bool
