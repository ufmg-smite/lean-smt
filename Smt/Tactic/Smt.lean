/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean

import Smt.Dsl.Sexp
import Smt.Reconstruct
import Smt.Reconstruct.Prop.Lemmas
import Smt.Translate.Query
import Smt.Preprocess
import Smt.Util

namespace Smt

open Lean hiding Command
open Elab Tactic Qq
open Smt Translate Query Reconstruct Util

def genUniqueFVarNames : MetaM (Std.HashMap FVarId String × Std.HashMap String FVarId) := do
  let lCtx ← getLCtx
  let st : NameSanitizerState := { options := {}}
  let (lCtx, _) := (lCtx.sanitizeNames st).run
  return lCtx.getFVarIds.foldl (init := ({}, {})) fun (m₁, m₂) fvarId =>
    let m₁ := m₁.insert fvarId (lCtx.getRoundtrippingUserName? fvarId).get!.toString
    let m₂ := m₂.insert (lCtx.getRoundtrippingUserName? fvarId).get!.toString fvarId
    (m₁, m₂)

def prepareSmtQuery (hs : List Expr) (goalType : Expr) (fvNames : Std.HashMap FVarId String) : MetaM (List Command) := do
  let goalId ← Lean.mkFreshMVarId
  Lean.Meta.withLocalDeclD goalId.name (mkNot goalType) fun g =>
  Query.generateQuery g hs fvNames

def withProcessedHints (mv : MVarId) (hs : List Expr) (k : MVarId → List Expr → MetaM α): MetaM α :=
  go mv hs [] k
where
  go (mv : MVarId) (hs : List Expr) (fvs : List Expr) (k : MVarId → List Expr → MetaM α): MetaM α := do
    match hs with
    | [] => k mv fvs
    | h :: hs =>
      if h.isFVar || h.isConst then
        go mv hs (h :: fvs) k
      else
        let mv ← mv.assert (← mkFreshId) (← Meta.inferType h) h
        let ⟨fv, mv⟩ ← mv.intro1
        go mv hs (.fvar fv :: fvs) k

def smt (mv : MVarId) (hs : List Expr) (timeout : Option Nat := none) : MetaM (List MVarId) := mv.withContext do
  -- 1. Process the hints passed to the tactic.
  withProcessedHints mv hs fun mv hs => mv.withContext do
  let (hs, mv) ← Preprocess.elimIff mv hs
  mv.withContext do
  let goalType : Q(Prop) ← mv.getType
  -- 2. Generate the SMT query.
  let (fvNames₁, fvNames₂) ← genUniqueFVarNames
  let cmds ← prepareSmtQuery hs (← mv.getType) fvNames₁
  let cmds := .setLogic "ALL" :: cmds
  trace[smt] "goal: {goalType}"
  trace[smt] "\nquery:\n{Command.cmdsAsQuery (cmds ++ [.checkSat])}"
  -- 3. Run the solver.
  let res ← solve (Command.cmdsAsQuery cmds) timeout
  -- 4. Print the result.
  -- trace[smt] "\nresult: {res}"
  match res with
  | .error e =>
    -- 4a. Print error reason.
    trace[smt] "\nerror reason:\n{repr e}\n"
    throwError "unable to prove goal, either it is false or you need to define more symbols with `smt [foo, bar]`"
  | .ok pf =>
    -- 4b. Reconstruct proof.
    let (p, hp, mvs) ← reconstructProof pf fvNames₂
    let mv ← mv.assert (← mkFreshId) p hp
    let ⟨_, mv⟩ ← mv.intro1
    let ts ← hs.mapM Meta.inferType
    let mut gs ← mv.apply (← Meta.mkAppOptM ``Prop.implies_of_not_and #[listExpr ts q(Prop), goalType])
    mv.withContext (gs.forM (·.assumption))
    return mvs

namespace Tactic

syntax smtHints := ("[" term,* "]")?
syntax smtTimeout := ("(" "timeout" " := " num ")")?

/-- `smt` converts the current goal into an SMT query and checks if it is
satisfiable. By default, `smt` generates the minimum valid SMT query needed to
assert the current goal. However, that is not always enough:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt
```
For the theorem above, `smt` generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(check-sat)
```
which is missing the hypotheses `hp` and `hpq` required to prove the theorem. To
pass hypotheses to the solver, use `smt [h₁, h₂, ..., hₙ]` syntax:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt [hp, hpq]
```
The tactic then generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(declare-const p Bool)
(assert p)
(assert (=> p q))
(check-sat)
```
-/
syntax (name := smt) "smt" smtHints smtTimeout : tactic

/-- Like `smt`, but just shows the query without invoking a solver. -/
syntax (name := smtShow) "smt_show" smtHints : tactic

def parseHints : TSyntax `smtHints → TacticM (List Expr)
  | `(smtHints| [ $[$hs],* ]) => hs.toList.mapM (fun h => elabTerm h.raw none)
  | `(smtHints| ) => return []
  | _ => throwUnsupportedSyntax

def parseTimeout : TSyntax `smtTimeout → TacticM (Option Nat)
  | `(smtTimeout| (timeout := $n)) => return some n.getNat
  | `(smtTimeout| ) => return some 5
  | _ => throwUnsupportedSyntax

@[tactic smt] def evalSmt : Tactic := fun stx => withMainContext do
  let mvs ← Smt.smt (← Tactic.getMainGoal) (← parseHints ⟨stx[1]⟩) (← parseTimeout ⟨stx[2]⟩)
  Tactic.replaceMainGoal mvs

@[tactic smtShow] def evalSmtShow : Tactic := fun stx => withMainContext do
  let g ← Meta.mkFreshExprMVar (← getMainTarget)
  let mv := g.mvarId!
  let hs ← parseHints ⟨stx[1]⟩
  withProcessedHints mv hs fun mv hs => mv.withContext do
  let (hs, mv) ← Preprocess.elimIff mv hs
  mv.withContext do
  let goalType ← mv.getType
  let cmds ← prepareSmtQuery hs (← mv.getType) (← genUniqueFVarNames).fst
  let cmds := cmds ++ [.checkSat]
  logInfo m!"goal: {goalType}\n\nquery:\n{Command.cmdsAsQuery cmds}"

end Smt.Tactic
