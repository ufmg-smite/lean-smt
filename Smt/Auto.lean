/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Auto.Tactic
import Smt.Tactic.Smt

open Lean in
def smtSolverFunc (ls : Array Auto.Lemma) (is : Array Auto.Lemma) : MetaM Expr := do
  let fi l := do
    let userName ← mkFreshUserName `inst
    let type ← Meta.mkAppM ``Inhabited #[l.type]
    let value ← Meta.mkAppOptM ``Inhabited.mk #[l.type, l.proof]
    return { userName, type, value }
  let is ← is.mapM fi
  let fl l := do
    let userName ← mkFreshUserName `h
    let type ← Auto.Lam2D.approxSimpNF l.type
    let value := l.proof
    return { userName, type, value }
  let ls ← ls.mapM fl
  let h ← Meta.mkFreshExprMVar (Expr.const ``False [])
  let (_, mv) ← h.mvarId!.assertHypotheses is
  let (fvs, mv) ← mv.assertHypotheses ls
  mv.withContext do
    let hs := fvs.map (.fvar ·)
    _ ← Smt.smt mv hs.toList
    -- Note: auto should allow solvers to export new goals to users
    -- for mv in mvs do
    --   logInfo m!"new : {mv}"
    instantiateMVars h

attribute [rebind Auto.Native.solverFunc] smtSolverFunc
