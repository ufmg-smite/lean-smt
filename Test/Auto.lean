import Mathlib.Algebra.Group.Defs
import Mathlib.Data.Set.Function

import Smt
import Smt.Auto

set_option auto.native true

example (x : α) : List.head? [x] = .some x := by
  have h₂ : ∀ (x : α) (ys : _), List.head? (x :: ys) = .some x := fun _ _ => rfl
  auto

example (x y : α) : [x] ++ [y] = [x, y] := by
  -- Invoke definition unfolding
  have h : ∀ (x y : List α), x ++ y = x.append y := fun _ _ => rfl
  auto [h] d[List.append]

variable [Group G]

theorem inverse' : ∀ (a : G), a * a⁻¹ = 1 := by
  auto [mul_assoc, one_mul, inv_mul_cancel]

theorem identity' : ∀ (a : G), a * 1 = a := by
  auto [mul_assoc, one_mul, inv_mul_cancel, inverse']

theorem unique_identity (e : G) : (∀ z, e * z = z) → e = 1 := by
  auto [mul_assoc, one_mul, inv_mul_cancel]

-- TODO: pre-process Nat away
-- example (x y : α) : List.get? [x, y] 1 = .some y := by
  -- auto d[List.get?]

set_option auto.mono.mode "fol"
set_option auto.mono.ciInstDefEq.mode "reducible"
set_option auto.mono.termLikeDefEq.mode "reducible"

variable [Nonempty α] [Nonempty β] {f : α → β} {s : Set α} {v u : Set β}

example : f '' s ⊆ v ↔ s ⊆ f ⁻¹' v := by
  auto [Set.subset_def, Set.mem_image, Set.mem_preimage]

example (h : Function.Injective f) : f ⁻¹' (f '' s) ⊆ s := by
  auto [Set.subset_def, Set.mem_preimage, Function.Injective.mem_set_image, h]

example : f '' (f ⁻¹' u) ⊆ u := by
  auto [Set.subset_def, Set.mem_image, Set.mem_preimage]

example (h : Function.Surjective f) : u ⊆ f '' (f ⁻¹' u) := by
  unfold Function.Surjective at h
  auto [Set.subset_def, Set.mem_image, Set.mem_preimage, h]
