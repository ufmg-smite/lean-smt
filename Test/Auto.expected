Test/Auto.lean:38:2: error: Auto.LamReif.reifTermCheckType :: LamTerm ((!9 (!13 (!1 !2)) !3) = (!0 (!1 !2) !3)) is not type correct
Test/Auto.lean:41:2: error: Auto.LamReif.reifTermCheckType :: LamTerm ((!6 (!11 (!1 (!2 !3))) !3) = (!0 (!1 (!2 !3)) !3)) is not type correct
Test/Auto.lean:44:2: error: Auto.LamReif.reifTermCheckType :: LamTerm ((!6 (!10 (!1 (!2 !3))) !3) = (!0 (!1 (!2 !3)) !3)) is not type correct
Test/Auto.lean:48:2: error: Auto.LamReif.reifTermCheckType :: LamTerm ((!6 (!10 !1) (!2 (!3 !1))) = (!0 !1 (!2 (!3 !1)))) is not type correct
